### DISCUSSION TITLE
why does juicefs have multiple storage layer?

### DISCUSSION THREAD

**Discussion by frankxieke** [juicedata/juicefs#323] [Q&A] [answered]
The storage format of one file in JuiceFS consists of three levels.

I am wondering why juicefs design these three levels? Could you give me a hint?

**Comment by frankxieke** [comment]
- [x] looking forward your reply

**Comment by xiaogaozi** [comment, marked as answer]
The short answer is: "chunk" is for speed-up seek big file, "slice" is for random write file, "block" is for speed-up read from and write to object storage.

**Reply by frankxieke** [reply]
thanks for your answer.  

but i am still wondering why "slice is variable size , and this is for rand write file?'


does it have any more details i can learn?

**Reply by davies** [reply]
A pwrite() can happen on anywhere on a file, and has any number of bytes, so a slice is used represent one or multiple continuous writes.
